### ISSUE TITLE
[Feature Request]: Add support for code cached ASAR mapping to snapshot for compiled secure execution

### ISSUE THREAD

**Issue by gpetrov** [electron/electron#33571]
### Preflight Checklist

- [X] I have read the [Contributing Guidelines](https://github.com/electron/electron/blob/main/CONTRIBUTING.md) for this project.
- [X] I agree to follow the [Code of Conduct](https://github.com/electron/electron/blob/main/CODE_OF_CONDUCT.md) that this project adheres to.
- [X] I have searched the [issue tracker](https://www.github.com/electron/electron/issues) for a feature request that matches the one I want to file, without success.

### Problem Description

Running secure safe code that can not be tampered has always been a challenge in Electron.
As all code is just included, even in ASAR packed form , it can be easily replaced or tampered with.

This also includes that the main.js file can be easily modified to remove a lot of restrictions like node integration, web and security and frame security etc.

### Proposed Solution

There is already a good support for generating snapshots with mksnapshot, however those snapshots run in very isolated space and are usually restricted by that.

Also those snapshots are just js code as well, so can be easily accessed as well.

When running js files chrome compiles those to bytecode on the fly for easier and faster access from the code cache.
Recently the Electron team added an API to set the code cache path https://www.electronjs.org/docs/latest/api/session#sessetcodecachepathpath

What if we can compile those js to code cache files before hand, bundle them in an ASAR and put this ASAR file in the snapshot?
So that the snapshot is accessed as virtual folder - like the regular ASAR files.

This will make Electron and its content much more secure and difficult to modify.

The snapshot ASAR loading can be further more secured by content validation check on startup when the snapshot is loaded, to verify that the snapshot integrity is correct and it has not being tampered. This can be done with public keys build in electrons binary.

### Alternatives Considered

Many alternatives has been considered, like using Bytenode https://github.com/bytenode/bytenode but as the Electron team is deprecating VM usage in the renderer those solutions will eventually cease to work. See https://github.com/electron/electron/pull/26087

**Comment by MarshallOfSound** [comment]
You're making a common mistake here and thinking about security incorrectly.

You're confusing the ability to read the apps code (which has no impact on security) with the ability to modify it.

Whether the code is JS, wasm, or bytecode it doesn't matter if you can read it or not.  The thing that some people want to protect against is application modification.

Electron supports this protection on platforms that support this kind of security posture.  Specifically, only macOS as it is the only OS that truly has runtime integrity / code signature enforcement.  Windows is getting there with applocker in enterprise deployments and as such it's possible electron will support this system on windows too one day.

The system I'm referring to is Electron Asar Integrity which is enforced via the Hardened Runtime on macOS and enabled via an electron fuse.  https://github.com/electron/fuses

TLDR: Just turn on Asar Integrity, it'll be on by default soon anyway.  And don't confuse code obfuscation with security, people will always be able to reverse engineer your client code. Always.

**Comment by gpetrov** [comment]
Yes @MarshallOfSound  you are right I'm mostly concerned about app code modifications. That is security to us as we don't want other people sneaking some malware in our app.

Good to know about the electron fuse and Asar integrity - but this is not a viable solution as it is Mac only. Is it so difficult to do an integrity check on all OS's during the start up?

Are you also saying that mapping the ASAR to electron snapshot is also a bad idea? Seems to me that the snapshot is a perfect candidate for memory mapped file like ASAR.

A good example of using modules in v8 snapshot is: https://github.com/cypress-io/v8-snapshot
and if we can have the whole asar as MMF from the snapshot it will be great

**Comment by MarshallOfSound** [comment]
> Are you also saying that mapping the ASAR to electron snapshot is also a bad idea?

I'm saying it has zero impact on either actual app security or the ability for people to read your code.  If it's all javascript, it's all visible in the debugger anyway with sufficient poking.

> Is it so difficult to do an integrity check on all OS's during the start up?

It's not that it's difficult, it's that IMO it's currently pointless.  In order for an app to be able to defend itself, it must be defended by the OS.  As an example, here is the trust chain on macOS.

* The OS validates the application on boot
* The OS validates the app as it is loaded into memory and executed
* The app uses a part of the validated binary to store some checksums and then validates the ASAR file

You can trust the ASAR _because_ you can trust the checksums _because_ you can trust the app in memory _because_ you can trust the OS.

On windows there is no such trust chain, even if the app validated the ASAR there is nothing on windows that validates the app.  If someone has write access to the ASAR they can just overwrite the app instead 🤷 Until windows introduces proper runtime binary validation (the closest thing they have is applocker which is off by default and enterprise only) there's very limited point implementing asar integrity on windows.

**Comment by gpetrov** [comment]
> I'm saying it has zero impact on either actual app security or the ability for people to read your code. If it's all javascript, it's all visible in the debugger anyway with sufficient poking.

Well bytecode, like the chrome code cache or the one produced by the vm code context - can be quite challenging to decode when the source is not included. That is what I'm also requesting - to have all the files in the ASAR already compiled to byte code and use them as ready to go code cache for chrome. Directly loading from the snapshot as memory mapped code cache folder.

That should already offer enough code tempering protection and indeed at least basic source code protection as the code isn't included.

> It's not that it's difficult, it's that IMO it's currently pointless. In order for an app to be able to defend itself, it must be defended by the OS. As an example, here is the trust chain on macOS.

Well this is definitely not pointless and you should never put all the validation checks only on OS level. When the application loader can verify itself and stop booting up when the validation doesn't pass - it is all enough protection. Of course additional OS protection will be fine, but that won't happen any time soon for Windows or Linux in the next 10 years. We better just make our own bullet proof loader and don't put the blame to the hands of others and wait doing nothing while malware spreads.

It is time to take responsibility and action.
